src/shogun/lib/any.h

```diff
@@ -178,13 +180,58 @@ namespace shogun
 		}
 
 		template <class T>
-		auto compare(T& lhs, T& rhs)
-		    -> decltype(compare_impl(maybe_most_important(), lhs, rhs))
+		inline bool compare(const T& lhs, const T& rhs)
 		{
 			return compare_impl(maybe_most_important(), lhs, rhs);
 		}
+
+		template <class T>
+		inline T clone_impl(general, T& value)
+		{
+			return value;
+		}
+
+		template <class T>
+		inline auto clone_impl(maybe_most_important, T* value)
+		    -> decltype(static_cast<void*>(value->clone()))
+		{
+			if (!value)
+				return nullptr;
+
+			return static_cast<void*>(value->clone());
+		}
+
+		template <class T>
+		inline auto clone(T& value)
+		    -> decltype(clone_impl(maybe_most_important(), value))
+		{
+			return clone_impl(maybe_most_important(), value);
+		}
+
+		template <class T>
+		inline T const* typed_pointer(const void* ptr)
+		{
+			return static_cast<T const*>(ptr);
+		}
+
+		template <class T>
+		inline const T& value_of(T const* ptr)
+		{
+			return *ptr;
+		}
+
+		template <class T>
+		inline T& mutable_value_of(void** ptr)
+		{
+			return *static_cast<T*>(*ptr);
+		}
 	}
 
+	using any_detail::typed_pointer;
+	using any_detail::value_of;
+	using any_detail::mutable_value_of;
```

Comment:
we are slowly getting quite complex with this stuff .... :D

Reply:
No that's actually to improve readability. I was having hard time checking all the * and &